packages/use-dataloader/src/useInfiniteDataLoader.ts

```diff
@@ -0,0 +1,227 @@
+    onSuccess,
+    keepPreviousData = false,
+    initialData,
+    dataLifetime,
+    pageKey,
+    params,
+    getNextPage,
+  } = config ?? {}
+  const requestRefs = useRef<DataLoader<ResultType, ErrorType>[]>([])
+  const [page, setPage] = useState(
+    params && pageKey ? params[pageKey] : undefined,
+  )
+  const nextPageRef = useRef(page)
+  const getNextPageRef = useRef(getNextPage)
+  const methodRef = useRef(() =>
+    method(
+      pageKey && page ? ({ ...params, [pageKey]: page } as ParamsType) : params,
+    ),
+  )
+  const paramsRef = useRef(
+    pageKey && page ? ({ ...params, [pageKey]: page } as ParamsType) : params,
+  )
+  const onSuccessRef = useRef(onSuccess)
+  const onErrorRef = useRef(onError ?? onGlobalError)
+  const [, setCounter] = useState(0)
+
+  const forceRerender = useCallback(() => {
+    setCounter(current => current + 1)
+  }, [])
+
+  const baseQueryKey = useMemo(
+    () => marshalQueryKey([marshalQueryKey(baseKey), page as string | number]),
+    [baseKey, page],
+  )
+
+  useEffect(() => {
+    requestRefs.current.forEach(request => request.removeObserver(() => null))
+    requestRefs.current = []
+    forceRerender()
+  }, [baseKey, forceRerender])
+
+  useEffect(
+    () => () => {
+      requestRefs.current.forEach(request => request.removeObserver(() => null))
+    },
+    [],
+  )
+
+  const getCurrentRequest = useCallback(() => {
+    const requestInRef = requestRefs.current.find(
+      request => request.key === baseQueryKey,
+    )
+    if (!requestInRef) {
+      const request = getOrAddRequest<ResultType, ErrorType>(baseQueryKey, {
+        enabled,
+        method: methodRef.current,
+      })
+      requestRefs.current.push(request)
+      request.addObserver(forceRerender)
+
+      return request
+    }
+
+    return requestInRef
+  }, [baseQueryKey, forceRerender, enabled, getOrAddRequest])
+
+  const request = getCurrentRequest()
+
+  const needLoad = useMemo(
+    () =>
+      !!(
+        enabled &&
+        (!request.dataUpdatedAt ||
+          !dataLifetime ||
+          (request.dataUpdatedAt &&
+            dataLifetime &&
+            request.dataUpdatedAt + dataLifetime < Date.now()))
+      ),
+    [enabled, request.dataUpdatedAt, dataLifetime],
+  )
+
+  const optimisticIsLoadingRef = useRef(needLoad)
+
+  const previousDataRef = useRef(request.data)
+
+  const isLoading = requestRefs.current.some(
+    req => req.status === StatusEnum.LOADING || optimisticIsLoadingRef.current,
+  )
+
+  // Current request is loading and its the first request added in the refs
+  const isLoadingFirstPage =
+    (request.status === StatusEnum.LOADING || optimisticIsLoadingRef.current) &&
+    requestRefs.current.length <= 1
+
+  const isSuccess = request.status === StatusEnum.SUCCESS
+
+  const isError = request.status === StatusEnum.ERROR
+
+  const isIdle = requestRefs.current.every(
+    req => req.status === StatusEnum.IDLE && !enabled,
+  )
+
+  const reload = useCallback(async () => {
```

Comment:
What happen when a page doesn't exists anymore ?

Reply:
data will be null/undefined/empty array and be filtered out when user get the data from the hook. May be if it doesnt fit some edge cases I can cont how many requests we have and try to get each one sequentially then by using getNextPage the hook can try to fill the same number of pages but if it cant it will not make useless calls ? :thinking: